### PR TITLE
samr21: add interaction between at86rf2xx radio state and CPU low power management

### PR DESCRIPTION
**Pull request by Hyungsin**
PM needs to keep tracking the radio state. Without this, MCU may go to the deepest sleep mode when the radio is on, resulting in a system crash. I observed that problem while trying to make the radio duty-cycle.

This PR adds interaction with PM layer to the at86rf2xx driver.

@miri64, you may review this too :)

### CODE REVIEW

**Comment by miri64** [comment]
Not an expert on power management or its API, so I'm reassigning. I can however imagine that the function `pm_radio_on()` gives some people headache, since there are other devices, too, that need to prevent sleep (and IIRC there is already some functionality in `pm` for that).

**Comment by miri64** [comment]
E.g. why can't you simply use `pm_block()`? The global state variable inside `pm_layered` seems to be a little bit not thought through. What if there are multiple radios e.g.? What if there are other devices that use this. Better to store this outside (e.g. in the flags of the device... `0x8000` is still free) or infer it by the device's state.

**Comment by Hyungsin** [comment]
Point is taken. 
I updated this PR so that it only modifies radio driver to support interaction with PM.

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 1]
```diff
@@ -486,6 +487,10 @@ uint8_t at86rf2xx_set_state(at86rf2xx_t *dev, uint8_t state)
         /* Go to SLEEP mode from TRX_OFF */
         gpio_set(dev->params.sleep_pin);
         dev->state = state;
+        /* Allow MCU to go to the full sleep mode */
+        if (old_state != AT86RF2XX_STATE_SLEEP) {
+            pm_unblock(PM_NUM_MODES-1);
```
Comment (miri64) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: Style comment: please leave spaces around operations.

**Comment by Hyungsin** [comment]
@miri64, fixed :)

**Comment by miri64** [comment]
Handing over remaining review to @kaspar030 or @haukepetersen :-)

**Comment by Hyungsin** [comment]
@gebart, I think your points are valid. 
The problematic board is SAMR21. I've not tested other platforms.
I will try to add this functionality on the cpu driver, not the radio driver :)

**Comment by kaspar030** [comment]
> Why not PM_NUM_MODES-2? I should test other power modes also :)

The question that needs answering is "why PM_NUM_MODES-2". This is certainly cpu dependent. Why is there a crash in the first place?

**Comment by Hyungsin** [comment]
@kaspar030, there are four power modes in SAMR21. [0: Standby, 1: IDLE_2, 2: IDLE_1, 3: IDLE_0]
FYI, PM_NUM_MODES = 3 for SAMR21. Without any explicit call of pm_block(), PM_LAYERED always tries to let CPU sleep with the Standby mode. The SAMR21 datasheet shows a user requirement as below:   
-------------------------------
STANDBY mode: All clock sources are stopped, except those where the RUNSTDBY bit is set. Regulator operates in low-power mode. Before entering standby mode the user must make sure that a significant amount of clocks and peripherals are disabled, so that the voltage regulator is not overloaded.
--------------------------------
So, at least for SAMR21, radio state should be reported to PM_LAYERED to mitigate any malfunction. 

Then the next question is what power mode should be blocked. I simply blocked IDLE_1 (PM_NUM_MODES-1) so that CPU goes to IDLE_0 in the idle thread. Given that IDLE_0 is the highest low power mode, there would be chances to use other power modes as well (IDLE_1 or IDLE_2). I can clarify this tomorrow in my lab :-)

**Comment by gebart** [comment]
@Hyungsin aren't you in fact blocking all PM by blocking PM_NUM_MODES - 1 (= 3: IDLE_0).

See https://github.com/RIOT-OS/RIOT/blob/master/sys/pm_layered/pm.c#L49
pm_set will be called with mode = PM_NUM_MODES 
https://github.com/RIOT-OS/RIOT/blob/master/cpu/samd21/periph/pm.c#L54

**Comment by Hyungsin** [comment]
@gebart , good point and I think it is kind of awkward. 

By definition, PM_NUM_MODES should be "the number of possible power modes". 
But in the current implementation, PM_NUM_MODES is "1" less than the number. 

See https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam0_common/include/periph_cpu_common.h#L66
PM_NUM_MODES = 3 (which is the highest low power mode, IDLE_0) 
So blocking PM_NUM_MODES-1 does not block all PM but enables IDLE_0.

**Comment by Hyungsin** [comment]
It has been a while since I found this problem and I believed that it still exists now.
But after some tests today, I think the current RIOT has no problem without reporting the radio state.  
I don't know what has been changed, but it is good to close this PR. 
@gebart, @kaspar030, @miri64, sorry for taking your time :(

**Pull request closed without merging**